File: scheduler.py
import datetime
import random
import copy

def get_week_number(date_obj):
    return date_obj.isocalendar()[:2]

def parse_dt(v):
    if isinstance(v, str): 
        try:
            return datetime.datetime.strptime(v, "%Y-%m-%d").date()
        except:
            return None
    return v

def calculate_points(historico, pessoas_keys, target_date=None):
    if target_date is None:
        target_date = datetime.date.today()
        
    pontos_preta = {p: 0 for p in pessoas_keys}
    pontos_vermelha = {p: 0 for p in pessoas_keys}
    recent_duties_count = {p: 0 for p in pessoas_keys}
    last_weekend_worked = {p: None for p in pessoas_keys}
    last_worked_date = {p: None for p in pessoas_keys}
    active_in_last_30_days = set()
    
    peso_guarda = 3
    peso_plantao = 2
    peso_apoio = 1
    
    historico_ordenado = sorted(historico, key=lambda x: parse_dt(x['data']))
    for reg in historico_ordenado:
        dt = parse_dt(reg['data'])
        if not dt: continue
        wkdy = dt.weekday()
        is_reg_vermelha = wkdy in [4, 5, 6]
        is_we = (wkdy == 5 or wkdy == 6)
        c_week = get_week_number(dt)
        
        days_diff = (target_date - dt).days
        is_recent = 0 < days_diff <= 7
        is_active = 0 < days_diff <= 30
        
        def process_duty(person_list, peso):
            for p in person_list:
                p = str(p)
                if p not in pessoas_keys: continue
                if is_active: active_in_last_30_days.add(p)
                
                if is_reg_vermelha: pontos_vermelha[p] += peso
                else: pontos_preta[p] += peso
                
                last_worked_date[p] = dt
                if is_we: last_weekend_worked[p] = c_week
                if is_recent: recent_duties_count[p] += 1
                
        process_duty(reg.get('guarda', []), peso_guarda)
        process_duty(reg.get('plantao', []), peso_plantao)
        process_duty(reg.get('apoio', []), peso_apoio)

    avg_preta = 0
    avg_vermelha = 0
    if active_in_last_30_days:
        avg_preta = sum(pontos_preta[p] for p in active_in_last_30_days) / len(active_in_last_30_days)
        avg_vermelha = sum(pontos_vermelha[p] for p in active_in_last_30_days) / len(active_in_last_30_days)
        
    for p in pessoas_keys:
        threshold_preta = max(0, avg_preta - peso_guarda)
        if pontos_preta[p] < threshold_preta:
            pontos_preta[p] = threshold_preta
            
        threshold_vermelha = max(0, avg_vermelha - peso_guarda)
        if pontos_vermelha[p] < threshold_vermelha:
            pontos_vermelha[p] = threshold_vermelha
            
    return pontos_preta, pontos_vermelha, recent_duties_count, last_weekend_worked, last_worked_date

def generate_daily_schedule(target_date, has_guarda, has_plantao, has_apoio, dispensas, current_state):
    pessoas_db = copy.deepcopy(current_state.get('pessoas', {}))
    historico = current_state.get('historico_escalas', [])
    
    pontos_preta, pontos_vermelha, recent_duties_count, last_weekend_worked, last_worked_date = calculate_points(historico, list(pessoas_db.keys()), target_date)

    weekday = target_date.weekday()
    is_meio_semana = weekday in [0, 1, 2, 3]

    available = []
    for p_str, p_data in pessoas_db.items():
        if not p_data.get('ativo', True):
            continue
            
        if p_data.get('is_po', False) and is_meio_semana:
            continue
            
        p_int = int(p_str)
        is_dispensado = False
        if p_int in dispensas:
            for d_start, d_end in dispensas[p_int]:
                if d_start <= target_date <= d_end:
                    is_dispensado = True
                    break
        if not is_dispensado:
            available.append(p_str)
            
    is_weekend = (weekday == 5 or weekday == 6)
    current_week = get_week_number(target_date)
    delta = datetime.timedelta(days=1)
    
    target_is_vermelha = weekday in [4, 5, 6]
    
    scores = {}
    for p in available:
        base_score = pontos_vermelha[p] if target_is_vermelha else pontos_preta[p]
        recent_penalty = recent_duties_count.get(p, 0) * 15
        consec_penalty = 0
        if last_worked_date.get(p) == target_date - delta:
            consec_penalty = 50
        scores[p] = base_score + recent_penalty + consec_penalty
    
    def filter_candidates(cands, block_consec_days, block_consec_weekends):
        valid = []
        for p in cands:
            if block_consec_days and last_worked_date.get(p) == target_date - delta:
                continue
            if block_consec_weekends and is_weekend:
                last_we = last_weekend_worked.get(p)
                if last_we:
                    ly, lw = last_we
                    cy, cw = current_week
                    if ly == cy and lw == cw - 1:
                        continue
                    elif ly == cy - 1 and lw == 52 and cw == 1:
                        continue
            valid.append(p)
        return valid

    req_guarda = 24 if has_guarda else 0
    req_plantao = 6 if has_plantao else 0
    req_apoio = 2 if has_apoio else 0
    total_req = req_guarda + req_plantao + req_apoio
    
    # Estratégia de Relaxamento de Regras
    candidates = filter_candidates(available, block_consec_days=True, block_consec_weekends=True)
    if len(candidates) < total_req:
        candidates = filter_candidates(available, block_consec_days=False, block_consec_weekends=True)
    if len(candidates) < total_req:
        candidates = filter_candidates(available, block_consec_days=False, block_consec_weekends=False)
    if len(candidates) < total_req:
        candidates = list(available)
        
    random.shuffle(candidates)
    candidates.sort(key=lambda x: scores[x])
    
    selected_guarda = []
    selected_plantao = []
    selected_apoio = []
    
    remaining_candidates = list(candidates)
    
    for p in list(remaining_candidates):
        if len(selected_guarda) >= req_guarda: break
        is_sargentiacao = pessoas_db.get(p, {}).get('is_sargentiacao', False)
        if is_sargentiacao and is_meio_semana: continue
        selected_guarda.append(p)
        remaining_candidates.remove(p)

    for p in list(remaining_candidates):
        if len(selected_plantao) >= req_plantao: break
        selected_plantao.append(p)
        remaining_candidates.remove(p)

    for p in list(remaining_candidates):
        if len(selected_apoio) >= req_apoio: break
        is_sargentiacao = pessoas_db.get(p, {}).get('is_sargentiacao', False)
        if is_sargentiacao and is_meio_semana: continue
        selected_apoio.append(p)
        remaining_candidates.remove(p)

    result_data = {
        'data': target_date.strftime("%Y-%m-%d"),
        'dia_semana': weekday,
        'guarda': selected_guarda,
        'plantao': selected_plantao,
        'apoio': selected_apoio,
        'has_guarda': has_guarda,
        'has_plantao': has_plantao,
        'has_apoio': has_apoio
    }
        
    new_state = copy.deepcopy(current_state)
    if 'historico_escalas' not in new_state:
        new_state['historico_escalas'] = []
    
    # Opcional: Adicionar na memória do preview, mas o app.py vai gerenciar a confirmação
    new_state['historico_escalas'].append(result_data)
        
    return result_data, new_state

File: GeradorEscalasWeb/backend/scheduler.go
package backend

import (
	"fmt"
	"math/rand"
	"sort"
	"strings"
	"time"
)

type ScoreData struct {
	PontosPreta           map[string]float64 `json:"pontos_preta"`
	PontosVermelha        map[string]float64 `json:"pontos_vermelha"`
	RecentDutiesCount     map[string]int     `json:"recent_duties_count"`
	LastWeekendWorked     map[string]int     `json:"last_weekend_worked"`
	LastWeekendWorkedYear map[string]int     `json:"last_weekend_worked_year"`
	LastWorkedDate        map[string]string  `json:"last_worked_date"`
}

func parseDt(dateStr string) (time.Time, error) {
	return time.Parse("2006-01-02", dateStr)
}

func getWeekNumber(dt time.Time) (int, int) {
	return dt.ISOWeek()
}

func CalculatePoints(historico []HistoricoEscala, pessoasKeys []string, targetDate time.Time, roleConfigs map[string]RoleConfig) ScoreData {
	sd := ScoreData{
		PontosPreta:           make(map[string]float64),
		PontosVermelha:        make(map[string]float64),
		RecentDutiesCount:     make(map[string]int),
		LastWeekendWorked:     make(map[string]int),
		LastWeekendWorkedYear: make(map[string]int),
		LastWorkedDate:        make(map[string]string),
	}

	activeInLast30Days := make(map[string]bool)

	for _, p := range pessoasKeys {
		sd.PontosPreta[p] = 0
		sd.PontosVermelha[p] = 0
		sd.RecentDutiesCount[p] = 0
	}

	sort.Slice(historico, func(i, j int) bool {
		dtI, _ := parseDt(historico[i].Data)
		dtJ, _ := parseDt(historico[j].Data)
		return dtI.Before(dtJ)
	})

	for _, reg := range historico {
		dt, err := parseDt(reg.Data)
		if err != nil {
			continue
		}

		wkdy := int(dt.Weekday()) // 0=Sun..6=Sat
		isRegVermelha := wkdy == 5 || wkdy == 6 || wkdy == 0
		isWe := wkdy == 6 || wkdy == 0

		year, cWeek := getWeekNumber(dt)
		daysDiff := int(targetDate.Sub(dt).Hours() / 24)
		isRecent := daysDiff > 0 && daysDiff <= 7
		isActive := daysDiff > 0 && daysDiff <= 30

		for roleName, personList := range reg.Escalados {
			peso := 1.0
			if cfg, ok := roleConfigs[roleName]; ok {
				peso = cfg.Weight
			}

			for _, p := range personList {
				if _, exists := sd.PontosPreta[p]; !exists {
					continue
				}
				if isActive {
					activeInLast30Days[p] = true
				}
				if isRegVermelha {
					sd.PontosVermelha[p] += peso
				} else {
					sd.PontosPreta[p] += peso
				}
				sd.LastWorkedDate[p] = dt.Format("2006-01-02")
				if isWe {
					sd.LastWeekendWorked[p] = cWeek
					sd.LastWeekendWorkedYear[p] = year
				}
				if isRecent {
					sd.RecentDutiesCount[p]++
				}
			}
		}
	}

	avgPreta, avgVermelha := 0.0, 0.0
	if len(activeInLast30Days) > 0 {
		sumPreta, sumVermelha := 0.0, 0.0
		for p := range activeInLast30Days {
			sumPreta += sd.PontosPreta[p]
			sumVermelha += sd.PontosVermelha[p]
		}
		avgPreta = sumPreta / float64(len(activeInLast30Days))
		avgVermelha = sumVermelha / float64(len(activeInLast30Days))
	}

	// Calculate max weight just for minimum threshold logic
	maxWeight := 0.0
	for _, rc := range roleConfigs {
		if rc.Weight > maxWeight {
			maxWeight = rc.Weight
		}
	}
	if maxWeight == 0 {
		maxWeight = 1.0
	}

	for _, p := range pessoasKeys {
		thresholdPreta := avgPreta - maxWeight
		if thresholdPreta < 0 {
			thresholdPreta = 0
		}
		if sd.PontosPreta[p] < thresholdPreta {
			sd.PontosPreta[p] = thresholdPreta
		}

		thresholdVermelha := avgVermelha - maxWeight
		if thresholdVermelha < 0 {
			thresholdVermelha = 0
		}
		if sd.PontosVermelha[p] < thresholdVermelha {
			sd.PontosVermelha[p] = thresholdVermelha
		}
	}

	return sd
}

type GenerateOpts struct {
	TargetDate   string
	EnabledRoles []string
}

func GenerateDailySchedule(opts GenerateOpts, currentState AppState) (HistoricoEscala, error) {
	targetDate, err := parseDt(opts.TargetDate)
	if err != nil {
		return HistoricoEscala{}, err
	}

	pessoasKeys := make([]string, 0, len(currentState.Pessoas))
	for k := range currentState.Pessoas {
		pessoasKeys = append(pessoasKeys, k)
	}

	sd := CalculatePoints(currentState.HistoricoEscalas, pessoasKeys, targetDate, currentState.RoleConfigs)

	wkdy := int(targetDate.Weekday())
	isMeioSemana := wkdy >= 1 && wkdy <= 4
	isWeekend := wkdy == 6 || wkdy == 0
	targetIsVermelha := wkdy == 5 || wkdy == 6 || wkdy == 0

	var available []string
	for pStr, pData := range currentState.Pessoas {
		if !pData.Ativo {
			continue
		}
		if pData.PostoGrad != "Soldado EP" && pData.PostoGrad != "Soldado EV" {
			continue
		}
		if pData.ApenasFimDeSemana && isMeioSemana {
			continue
		}
		if pData.ApenasSemana && targetIsVermelha {
			continue
		}

		isDispensado := false
		if dispensas, ok := currentState.Dispensas[pStr]; ok {
			for _, d := range dispensas {
				dStart, _ := parseDt(d.Inicio)
				dEnd, _ := parseDt(d.Fim)
				if (targetDate.After(dStart) || targetDate.Equal(dStart)) && (targetDate.Before(dEnd) || targetDate.Equal(dEnd)) {
					isDispensado = true
					break
				}
			}
		}
		if !isDispensado {
			available = append(available, pStr)
		}
	}

	currYear, currWeek := getWeekNumber(targetDate)
	delta := 24 * time.Hour

	scores := make(map[string]float64)
	for _, p := range available {
		baseScore := sd.PontosPreta[p]
		if targetIsVermelha {
			baseScore = sd.PontosVermelha[p]
		}
		recentPenalty := float64(sd.RecentDutiesCount[p] * 15)
		consecPenalty := 0.0

		lastWorkedStr, hasWorked := sd.LastWorkedDate[p]
		if hasWorked && lastWorkedStr == targetDate.Add(-delta).Format("2006-01-02") {
			consecPenalty = 50.0
		}

		scoreVal := baseScore + recentPenalty + consecPenalty

		pData := currentState.Pessoas[p]
		if pData.ApenasSemana && pData.ApenasFimDeSemana {
			scoreVal -= 10000
		}
		if p == "349" || p == "344" {
			scoreVal -= 4
		}

		scores[p] = scoreVal
	}

	filterCandidates := func(cands []string, blockConsecDays, blockConsecWeekends bool) []string {
		var valid []string
		for _, p := range cands {
			if blockConsecDays {
				lastWorkedStr, hasWorked := sd.LastWorkedDate[p]
				if hasWorked && lastWorkedStr == targetDate.Add(-delta).Format("2006-01-02") {
					continue
				}
			}
			if blockConsecWeekends && isWeekend {
				lastWeYear, yOk := sd.LastWeekendWorkedYear[p]
				lastWeWeek, wOk := sd.LastWeekendWorked[p]
				if yOk && wOk {
					if lastWeYear == currYear && lastWeWeek == currWeek-1 {
						continue
					} else if lastWeYear == currYear-1 && lastWeWeek == 52 && currWeek == 1 {
						continue
					}
				}
			}
			valid = append(valid, p)
		}
		return valid
	}

	// Order roles by weight descending so hardest roles get assigned first
	rolesToAssign := append([]string{}, opts.EnabledRoles...)
	sort.Slice(rolesToAssign, func(i, j int) bool {
		wI := currentState.RoleConfigs[rolesToAssign[i]].Weight
		wJ := currentState.RoleConfigs[rolesToAssign[j]].Weight
		return wI > wJ
	})

	totalReq := 0
	for _, role := range rolesToAssign {
		totalReq += currentState.RoleConfigs[role].Required
	}

	candidates := filterCandidates(available, true, true)
	if len(candidates) < totalReq {
		candidates = filterCandidates(available, false, true)
	}
	if len(candidates) < totalReq {
		candidates = filterCandidates(available, false, false)
	}
	if len(candidates) < totalReq {
		candidates = append([]string{}, available...)
	}

	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	r.Shuffle(len(candidates), func(i, j int) {
		candidates[i], candidates[j] = candidates[j], candidates[i]
	})

	sort.SliceStable(candidates, func(i, j int) bool {
		return scores[candidates[i]] < scores[candidates[j]]
	})

	escaladosMap := make(map[string][]string)
	remainingCandidates := append([]string{}, candidates...)

	for _, roleName := range rolesToAssign {
		req := currentState.RoleConfigs[roleName].Required
		aptosList := currentState.RoleConfigs[roleName].Aptos
		aptosMap := make(map[string]bool)
		for _, a := range aptosList {
			aptosMap[a] = true
		}

		var selected []string

		for i := 0; i < len(remainingCandidates); {
			p := remainingCandidates[i]
			if len(selected) >= req {
				break
			}
			pData := currentState.Pessoas[p]

			// Role Aptitude Filtering
			if len(aptosList) == 0 {
				// According to new rules, if empty, NO ONE is apt
				i++
				continue
			}
			if !aptosMap[p] {
				i++
				continue
			}

			if pData.ApenasSemana {
				lowerRole := strings.ToLower(roleName)
				if !strings.Contains(lowerRole, "plant") {
					i++
					continue
				}
			}
			selected = append(selected, p)
			remainingCandidates = append(remainingCandidates[:i], remainingCandidates[i+1:]...)
		}

		if len(selected) < req {
			return HistoricoEscala{}, fmt.Errorf("Não há militares aptos suficientes para a função '%s'. Requisitado: %d, Alocados: %d. Por favor, adicione mais militares aptos na aba Configurações.", roleName, req, len(selected))
		}

		escaladosMap[roleName] = selected
	}

	diaSemanaPython := wkdy - 1
	if diaSemanaPython == -1 {
		diaSemanaPython = 6
	}

	return HistoricoEscala{
		Data:          opts.TargetDate,
		DiaSemana:     diaSemanaPython,
		Escalados:     escaladosMap,
		SemExpediente: targetIsVermelha,
		GuardaComp:    make(map[string]string),
	}, nil
}
